File: core/src/test/java/com/xmartlabs/bigbang/core/GeneralErrorHelperTest.java
package com.xmartlabs.bigbang.core;

import com.tspoon.traceur.Traceur;
import com.tspoon.traceur.TraceurConfig;
import com.xmartlabs.bigbang.core.helper.GeneralErrorHelper;

import org.junit.Test;

import java.security.GeneralSecurityException;

import io.reactivex.Observable;

import static org.hamcrest.CoreMatchers.equalTo;
import static org.hamcrest.MatcherAssert.assertThat;

public class GeneralErrorHelperTest {
  private Integer calls;

  private String generateException() throws GeneralSecurityException {
    throw new GeneralSecurityException("Something went wrong");
  }

  private void initializeLogger() {
    calls = 0;
    GeneralErrorHelper generalErrorHelper = new GeneralErrorHelper();
    generalErrorHelper.setErrorHandlerForThrowable(GeneralSecurityException.class, throwable -> calls++);
    TraceurConfig config = new TraceurConfig(
        true,
        Traceur.AssemblyLogLevel.NONE,
        generalErrorHelper.getGeneralErrorAction()::accept);
    Traceur.enableLogging(config);
  }

  @Test
  public void multipleTimesSameException() {
    initializeLogger();
    Observable.fromCallable(this::generateException)
        .map(String::trim)
        .map(String::trim)
        .subscribe();
    assertThat(calls, equalTo(1));
  }

  @Test
  public void multipleTimesDifferentExceptions() {
    initializeLogger();
    Observable.fromCallable(this::generateException)
        .map(String::trim)
        .map(String::trim)
        .subscribe();
    Observable.fromCallable(this::generateException)
        .map(String::trim)
        .map(String::trim)
        .subscribe();
    assertThat(calls, equalTo(2));
  }
}

File: core/src/test/java/com/xmartlabs/bigbang/core/GeneralErrorHelperTest.kt
package com.xmartlabs.bigbang.core

import com.tspoon.traceur.Traceur
import com.tspoon.traceur.TraceurConfig
import com.xmartlabs.bigbang.core.helper.GeneralErrorHelper
import io.reactivex.Observable
import org.hamcrest.CoreMatchers.equalTo
import org.hamcrest.MatcherAssert.assertThat
import org.junit.Test
import java.security.GeneralSecurityException

class GeneralErrorHelperTest {
  private var calls: Int = 0

  @Throws(GeneralSecurityException::class)
  private fun generateException(): String {
    throw GeneralSecurityException("Something went wrong")
  }

  private fun initializeLogger() {
    calls = 0
    val generalErrorHelper = GeneralErrorHelper()
    generalErrorHelper.setErrorHandlerForThrowable(GeneralSecurityException::class) { calls++ }
    val config = TraceurConfig(
        true,
        Traceur.AssemblyLogLevel.NONE,
        generalErrorHelper.generalErrorAction)
    Traceur.enableLogging(config)
  }

  @Test
  fun multipleTimesSameException() {
    initializeLogger()
    Observable.fromCallable { this.generateException() }
        .map { it.trim { it <= ' ' } }
        .map { it.trim { it <= ' ' } }
        .subscribe()
    assertThat<Int>(calls, equalTo(1))
  }

  @Test
  fun multipleTimesDifferentExceptions() {
    initializeLogger()
    Observable.fromCallable { this.generateException() }
        .map { it.trim { it <= ' ' } }
        .map { it.trim { it <= ' ' } }
        .subscribe()
    Observable.fromCallable { this.generateException() }
        .map { it.trim { it <= ' ' } }
        .map { it.trim { it <= ' ' } }
        .subscribe()
    assertThat<Int>(calls, equalTo(2))
  }
}
